### PR TITLE
Move most of Psych::Parser#parse in Ruby

### PR DESCRIPTION
**Pull request by eregon**
WIP but reviews welcome.
* A couple open questions:
  * Is it worth supporting the streaming reader (`newStreamReader()`)? That ones makes a method lookup and slow call for every byte read.
  * Is the logic for encoding other than UTF-8 ever needed/tested?

I guess I should remain as close as the original semantics in the big diff commit, and then maybe simplify in later commits.

### CODE REVIEW

File: lib/truffle/truffle/psych.rb [hunk 1]
```diff
@@ -17,30 +17,95 @@ def self.libyaml_version
...
+          else
+            raise e
+          end
+        end
+
+        if event?(event, :StreamStart)
+          @handler.start_stream(ANY)
+        elsif event?(event, :DocumentStart)
+          tags = []
+          version, explicit = doc_start_info(event) { |key, value|
```
Comment (nirvdrum) [inline lib/truffle/truffle/psych.rb]: I think we've settled on using `do...end` for multi-line blocks.

File: src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java [hunk 2]
```diff
@@ -90,442 +73,224 @@
...
+        } else if (flowStyle) {
+            return STYLE_FLOW;
+        } else {
+            return STYLE_NOT_FLOW;
         }
+    }
 
-        public static void throwException(final Throwable e) {
-            throwsUnchecked(e);
+    @TruffleBoundary
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java]: Do you need a boundary here? `stringFor` already has one.

File: lib/truffle/truffle/psych.rb (hunk 1 above)
Comment (nirvdrum) [reply lib/truffle/truffle/psych.rb]: It looks like there's a cop for this: http://rubocop.readthedocs.io/en/latest/cops_style/#styleblockdelimiters

I can do up a PR to enable that and clean up whatever code we have violating the style currently.

Comment (eregon) [reply lib/truffle/truffle/psych.rb]: :+1: I think that would help to keep it consistent.

File: src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java (hunk 2 above)
Comment (eregon) [reply src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java]: No, indeed.

**Comment by chrisseaton** [comment]
I'm not sure if it's better to use Psych differently or to try to use the C extension like MRI does?

**Comment by eregon** [comment]
Seems like a bigger effort to support the C ext at this point.
The interface is quite small (Parser#parse + 6 callbacks), but this was very old legacy code so it's a big diff.

**Comment by chrisseaton** [comment]
I mean switching to the streaming interface?

**Comment by eregon** [comment]
I ended up keeping the streaming support as it was there originally in this PR.

**Pull request merged**